File: 프로그래머스/unrated/142085. 디펜스 게임/디펜스 게임.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int n, int k, vector<int> enemy) {
    int answer = 0;
    
    priority_queue<int> pq;
    
    for(const auto& e : enemy){
        pq.push(e);
        if(k > 0 && n < e){
            n += pq.top();
            pq.pop();
            k--;
        }
        n -= e;
        if(n < 0){
            break;
        }
        answer++;
    }
    
    return answer;
}

File: 프로그래머스/unrated/142085. 디펜스 게임/디펜스 게임.java
import java.util.*;
class Solution {
    public int solution(int n, int k, int[] enemy) {
        int answer = 0;
        
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());     
        for(int e : enemy){
            pq.add(e);
            if(k > 0 && n < e){
                n += pq.poll();
                k--;
            }
            n -= e;
            if(n < 0){
                break;
            }
            answer++;
        }
        
        return answer;
    }
}